File: Solutions/0687/0687.py
# -*- coding:utf-8 -*-
# Author: washing
# DateTime: 2022/9/2 10:18
# File: 0687.py
# Desc: 


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        self.ret = 0
        def parse(root):
            nlv = None
            nrv = None
            nls = 0
            nrs = 0
            tls = 0
            trs = 0
            if root.left:
                nlv, nls = parse(root.left)
            if root.right:
                nrv, nrs = parse(root.right)
            if root.val == nlv:
                tls = nls+1
            if root.val == nrv:
                trs += nrs+1
            self.ret = max(self.ret, tls+trs)
            return root.val, max(tls, trs)
        if root: parse(root)
        return self.ret

File: Solutions/0687/0687.c
/***************************************************************************************************
 * Copyright: washing
 * FileName: 0687.c
 * Author: washing
 * Version: 1.0
 * Date: 2022-9-2 10:31:57
***************************************************************************************************/


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


int longestUnivaluePath(struct TreeNode* root){
    int ret = 0;
    int parse(struct TreeNode* root){
        int nls = 0, nrs = 0, tls = 0, trs = 0;
        if (root->left){
            nls = parse(root->left);
            if (root->val == root->left->val) tls = nls+1;
        }
        if (root->right){
            nrs = parse(root->right);
            if (root->val == root->right->val) trs = nrs+1;
        }
        ret = ret > tls+trs ? ret : trs+tls;
        return tls > trs ? tls : trs;
    }
    if (root) parse(root);
    return ret;
}
